clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -66,143 +67,206 @@ class device_vector {
   }
 
   template <typename OtherAllocator>
-  device_vector(const device_vector<T, OtherAllocator> &v)
-      : _alloc(get_default_queue()), _storage(v.real_begin()), _size(v.size()),
-        _capacity(v.capacity()) {}
+  device_vector(const device_vector<T, OtherAllocator> &other,
+                const Allocator &alloc)
+      : _alloc(alloc) {
+    _size = other.size();
+    _capacity = other.capacity();
+    _storage = ::std::allocator_traits<Allocator>::allocate(_alloc, _capacity);
+    _construct(other.begin(), other.end());
+  }
+
+  template <typename OtherAllocator>
+  device_vector(const device_vector<T, OtherAllocator> &other)
+      : device_vector(other, get_default_queue()) {}
 
   template <typename OtherAllocator>
   device_vector(::std::vector<T, OtherAllocator> &v)
-      : _alloc(get_default_queue()), _size(v.size()) {
-    _set_capacity_and_alloc();
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                v.begin(), v.end(), this->begin());
-    }
-  }
+      : device_vector(v.begin(), v.end()) {}
+
+  template <typename OtherAllocator>
+  device_vector(::std::vector<T, OtherAllocator> &v, const Allocator &alloc)
+      : device_vector(v.begin(), v.end(), alloc) {}
 
   template <typename OtherAllocator>
   device_vector &operator=(const ::std::vector<T, OtherAllocator> &v) {
     resize(v.size());
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
+    ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
                 v.begin(), v.end(), begin());
-    }
     return *this;
   }
+
   device_vector &operator=(const device_vector &other) {
     // Copy assignment operator:
+    if constexpr (::std::allocator_traits<Allocator>::
+                      propagate_on_container_copy_assignment::value) {
+      clear();
+      ::std::allocator_traits<Allocator>::deallocate(_alloc, _storage,
+                                                     _capacity);
+      _capacity = 0;
+      _alloc = other._alloc;
+    }
+    reserve(other.capacity());
     resize(other.size());
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
+    ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
                 other.begin(), other.end(), begin());
-    }
     return *this;
   }
   device_vector &operator=(device_vector &&other) {
     // Move assignment operator:
-    device_vector dummy(::std::move(other));
-    this->swap(dummy);
+    if constexpr (::std::allocator_traits<Allocator>::
+                      propagate_on_container_move_assignment::value) {
+      clear();
+      ::std::allocator_traits<Allocator>::deallocate(_alloc, _storage,
+                                                     _capacity);
+      _alloc = std::move(other._alloc);
```

Comment:
`std::` -> `::std::` if we want to be consistent with the convention we use.

Reply:
Thanks, fixed this and others in the PR.